### ISSUE TITLE
storage/mysql: Remove `initial_gtid_set` such that ALTER SOURCE can be performed after source-creation GTID set is gone

### ISSUE THREAD

**Issue by rjobanp** [MaterializeInc/materialize#27468]
### Feature request

https://github.com/MaterializeInc/materialize/pull/25054 introduced the `initial_gtid_set` field on MySQL sources to ensure we always snapshot from a consistent point across tables. When a user wants to add a new subsource long after the source was created, they will be unable to if this `initial_gtid_set` no longer exists in the replication binlog. We should find a way to remove this constraint while preserving the correctness property fixed by that PR.

**Comment by petrosagg** [comment]
Quick sketch of what I think the solution looks like:

When the storage controller creates the remap collection shard for an ingestion it checks whether it has been initialized (i.e if upper != {T_min}). If it has not initialized then it talks to the upstream system to acquire a read hold. That read hold will have an associated frontier `F_readhold`, which is the set of times for which we can correctly accumulate the contents of the upstream system. Using a source specific function we take this `F_readhold` frontier and step it forward to convert it into an upper frontier `U_init` with the property that `{!(U_init <= t) for t in F_readhold}`. The controller then writes the update `(U_init, T_min, +1)` to the remap collection and advances the upper to `now()`. Alternatively, if the upstream system can directly provide an upper frontier that is "strictly beyond" the frontier of our readhold we can use that instead of stepping the read hold frontier forward. In both cases what we end up with is a remap collection whose since frontier can be used to track the minimum times at which we can read the source.

When the time comes to render an ingestion dataflow we will be doing that as of some frontier. That as_of frontier will be expressed in MZ times but it can be directly translated into a source upper frontier, which can be used to figure out which updates of the upstream source belong to the snapshot and which ones need to be emitted as standalone updates.

Furthermore, as the since frontier of the remap collection advances we are effectively requiring fewer and fewer distinct times to be available upstream. This process of downgrading the remap collection is what should drive the downgrade of the read hold upstream. This is in contrast to what we do today where the upstream read hold is downgraded eagerly as soon as we make any progress with our upper, which makes it tricky to snapshot newly added subsources correctly.

**Comment by petrosagg** [comment]
> So we could remove this constraint and maintain parity with our postgres source, at the cost of relaxing the 'correctness' property and allowing the user the workflow of adding a new subsource a long after the initial source was created.

Just to clarify, the constraint that the initial_gtid_set of a subsource should not change cannot be lifted without risking correctness issues much worse than a subsource appearing empty (it can cause double ingestion of updates). What we can do instead of the non-trivial amount of work described above is record an initial_gtid_set per subsource which also has the property that once set it cannot change.

**Comment by rjobanp** [comment]
Ah yes, I misunderstood, thanks. I think that plan `record an initial_gtid_set per subsource` seems to be the best tradeoff between implementation-time and acceptable workflow for users so I will proceed with that unless I hear otherwise

**Comment by benesch** [comment]
> I think that plan `record an initial_gtid_set per subsource` seems to be the best tradeoff between implementation-time and acceptable workflow for users so I will proceed with that unless I hear otherwise

Agreed. No need to make the problem harder on ourselves by offering more consistency than the PostgreSQL source currently offers.

We do need to fix correctness property #2 for subsources at some point, but we can do that for both PostgreSQL and MySQL at some future point.

**Comment by rjobanp** [comment]
@petrosagg I'm about halfway through implementing this but am realizing that a separate `initial_gtid_set` per subsource seems to defeat the whole purpose of this field to begin with. The initial correctness issue this field was meant to fix is described here: https://github.com/MaterializeInc/materialize/issues/24982 

The only way this seems to provide any benefit if done on a per-subsource basis is if a snapshot can be halfway committed for a subsource, which isn't possible. Or if updates were emitted for a subsource before the snapshot completed and then a restart occurred, restarting the snapshot. However in that case the source upper wouldn't have advanced beyond the minimum frontier so those other emitted updates shouldn't have been committed.

> Just to clarify, the constraint that the initial_gtid_set of a subsource should not change cannot be lifted without risking correctness issues much worse than a subsource appearing empty (it can cause double ingestion of updates).

Basically - can you explain how this could actually happen?

**Comment by rjobanp** [comment]
Or is it just the fact that we know this value will still be the same for all subsources created together, such that we do preserve the correctness among the subsources created during the initial CREATE SOURCE, and any other subsources added later will each be independently correct from the point of their initial_gtid_set and beyond?

**Comment by rjobanp** [comment]
Ah, I think I may understand now! We are using this field to decide where to start the mysql replication stream when a subsource's resume-upper is at the minimum frontier. If we didn't record an `initial_gtid_set` that we know is at least before the `snapshot_gtid_set`, we wouldn't know where to start the replication stream from to ensure overlap with the data present in the snapshot (and emit rewinds for any overlap). If we happened to start the replication stream from a point after the snapshot_gtid_set, we would unknowingly miss any transactions that happened in-between.

**Comment by bosconi** [comment]
@rjobanp @petrosagg this might require controller work in the end? I've tagged it with a new `A-controller` label (which we will use to get a global view of controller-involved work). Please remove if this is not the case.
